### PR TITLE
repart: Allow configuring sector size

### PR DESCRIPTION
**Pull request by daandemeyer**
Let's allow users to configure the logical and physical sector size of their image. This is required when building images for a 4k sector size disk on a 512b sector size host or vice-versa.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, so what precisely is the difference between physical and logical sector size? I don't get it?

**Comment by DaanDeMeyer** [comment]
> hmm, so what precisely is the difference between physical and logical sector size? I don't get it?

Logical is what the device reports to the OS. Physical is what it actually uses internally. For fdisk, this matters for alignment. e.g. the first LBA depends only on the logical sector size, but partitions are aligned to the physical sector size to ensure performant I/O. But after digging through libfdisk again, you don't necessarily need to be able to configure the physical sector size, you can do the same by configuring the grain, so we should probably just expose that and keep the physical sector size hidden. Will change PR

**Comment by poettering** [comment]
hmm, but so far the idea was to *always* align everything to 4K to any such mess?

why allow aligning to less than 4k?

**Comment by poettering** [comment]
hmm, or is this PR about what to pass down to the mkfs tools?

i guess yeah, for that it might make sense to make things configurable.

so currently we ignored all the physical/logical/grain stuff of libfdisk, and instead always operated on bytes level, and aligned on our own to 4K and gave libfdisk the properly aligned values. And I think that makes a ton of sense, and we should stick to that. (we kinda have to, since the "elastic" size calculations we support would otherwise not work)

but yeah, it might make sense to make configurable what we pass to the file systems, for file systems that allow configuration of the intended sector size.

**Comment by bluca** [comment]
Yeah I'll need this too - fun fact, in order to be able to use fsverity the filesystem block size MUST be equal to the memory page size. So I'd even drop the 4k high limit, given it's not unlikely on arm64 to use larger pages.

**Comment by bluca** [comment]
> Yeah I'll need this too - fun fact, in order to be able to use fsverity the filesystem block size MUST be equal to the memory page size. So I'd even drop the 4k high limit, given it's not unlikely on arm64 to use larger pages.

https://www.kernel.org/doc/html/latest/filesystems/fsverity.html#ext4

> Currently, ext4 verity only supports the case where the Merkle tree block size, filesystem block size, and page size are all the same.

**Comment by DaanDeMeyer** [comment]
> so currently we ignored all the physical/logical/grain stuff of libfdisk, and instead always operated on bytes level, and aligned on our own to 4K and gave libfdisk the properly aligned values. And I think that makes a ton of sense, and we should stick to that. (we kinda have to, since the "elastic" size calculations we support would otherwise not work)

The thing is, if I understand correctly, probing of the GPT partition table happens at the first LBA, which is dependent on the sector size of the disk. For regular files, libfdisk defaults to 512 as the sector size but if you try to use this on a disk with 4K sector size, it'll try to start reading the GPT header at offset 4K which will obviously result in garbage. So we need a way to configure the sector size to avoid this problem.

It very much sucks that the format is defined in terms of the sector size which is disk dependent since it means you can't make disk images that are sector size independent but I'm not sure what we can do about that.

**Comment by DaanDeMeyer** [comment]
The filesystem defaults are based on the running kernel which don't work anymore if we start using a different sector size so we override their defaults to the right value. In case of ext4, the block size just needs to exceed the sector size so we default to 4096 to make sure the block size always exceeds the sector size.

**Comment by poettering** [comment]
> The thing is, if I understand correctly, probing of the GPT partition table happens at the first LBA, which is dependent on the sector size of the disk. For regular files, libfdisk defaults to 512 as the sector size but if you try to use this on a disk with 4K sector size, it'll try to start reading the GPT header at offset 4K which will obviously result in garbage. So we need a way to configure the sector size to avoid this problem.

Are you sure that it cannot auto-detect the 

> The thing is, if I understand correctly, probing of the GPT partition table happens at the first LBA, which is dependent on the sector size of the disk. For regular files, libfdisk defaults to 512 as the sector size but if you try to use this on a disk with 4K sector size, it'll try to start reading the GPT header at offset 4K which will obviously result in garbage. So we need a way to configure the sector size to avoid this problem.

Are you sure that libfdisk is unable to handle cases where the block device's sector size doesn't match the one in the GPT header? at least the block device being 512 with an image prepped for 4K should be fine, no? @karelzak what's the status there?

If this really is an issue, then I guess we should tweak our gpt dissector to auto-detect the intended sector size of the GPT disk image before we attach it to a loopback device, and then set the sector size of the loopback device to the one the gpt partition table declares. auto-detecting the sector size from the image should be easy, since the gpt header is either placed at bytes 512 or at byte 4k and should have the right gpt signature plus the right LBA offset stored in it.

@karelzak does libfdisk/libblkid have an API to query the intended block size of a disk image from the GPT header (or any other partition table header)? we could roll our own code for that, but ideally we'd have that exposed in one (or both) of these two libs?

File: src/partition/repart.c [hunk 1]
```diff
@@ -1973,16 +1974,20 @@ static int context_load_partition_table(Context *context) {
         assert(context->end == UINT64_MAX);
         assert(context->total == UINT64_MAX);
 
+        c = fdisk_new_context();
+        if (!c)
+                return log_oom();
+
+        r = fdisk_save_user_sector_size(c, /* phy= */ 0, arg_sector_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set sector size: %m");
```
Comment (poettering) [inline src/partition/repart.c:1983]: unless I am mistaken `r` is not a real error code here, but libfdisk just returns -1 on failure always... maybe pass `SYNTHETIC_ERRNO(EIO)` as first arg here, for lack of a better error code to return

File: src/partition/repart.c [hunk 2]
```diff
@@ -1973,16 +1974,20 @@ static int context_load_partition_table(Context *context) {
         assert(context->end == UINT64_MAX);
         assert(context->total == UINT64_MAX);
 
+        c = fdisk_new_context();
+        if (!c)
+                return log_oom();
+
+        r = fdisk_save_user_sector_size(c, /* phy= */ 0, arg_sector_size);
```
Comment (poettering) [inline src/partition/repart.c]: if arg_sector_size is zero we probably should skip this?

File: src/shared/mkfs-util.c [hunk 3]
```diff
@@ -487,15 +490,35 @@ int make_filesystem(
                                 return log_oom();
                 }
 
-        } else if (streq(fstype, "vfat"))
+                if (sector_size > 0) {
+                        r = strv_extend(&argv, "-s");
+                        if (r < 0)
+                                return log_oom();
+
```
Comment (poettering) [inline src/shared/mkfs-util.c:478]: maybe don't bother storing the ret val in `r` if you then propagate log_oom() anyway.

i.e. make this:

```c
if (strv_extend(…) < 0)
        return log_oom();
```

here and everywhere below similar

File: src/partition/repart.c [hunk 4]
```diff
@@ -1973,16 +1974,20 @@ static int context_load_partition_table(Context *context) {
...
-                if (!c)
-                        return log_oom();
-
-                r = fdisk_assign_device(c, context->node, arg_dry_run);
-        } else
-                r = fdisk_new_context_fd(context->backing_fd, arg_dry_run, &c);
+        r = fdisk_assign_device(
+                c,
+                context->backing_fd >= 0 ? FORMAT_PROC_FD_PATH(context->backing_fd) : context->node,
+                arg_dry_run);
```
Comment (poettering) [inline src/partition/repart.c]: please indent parameter lists to 16ch, no 8ch

**Comment by poettering** [comment]
lgtm, just some minor stuff.

**Comment by poettering** [comment]
once https://github.com/systemd/systemd/pull/26092 landed this configured sector stuff should also be propagated into any loopback device we create

**Comment by DaanDeMeyer** [comment]
Adressed comments, PTAL!

File: src/basic/parse-util.c [hunk 5]
```diff
@@ -256,6 +256,27 @@ int parse_size(const char *t, uint64_t base, uint64_t *size) {
...
+        if (ss < 512 || ss > 4096) /* Allow up to 4K due to dm-crypt support and 4K alignment by the homed LUKS backend */
+                return log_error_errno(SYNTHETIC_ERRNO(ERANGE), "Sector size not between 512 and 4096: %s", t);
+        if (!ISPOWEROF2(ss))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Sector size not power of 2: %s", t);
+
+        *ret = ss;
+        return 0;
+}
+
+
```
Comment (poettering) [inline src/basic/parse-util.c]: double empty line

File: man/systemd-repart.xml [hunk 6]
```diff
@@ -387,6 +387,15 @@
         <command>systemd-repart</command> was executed.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--sector-size=</option><arg>BYTES</arg></term>
```
Comment (medhefgo) [inline man/systemd-repart.xml:391]: Any reason why not support si units? `--sector-size=4K`

Comment (daandemeyer) [reply man/systemd-repart.xml:391]: Lazyness, `parse_sector_size()` didn't support it. Given that the only possible values are 512, 2048 and 4096, the added benefit seems rather low

**Comment by bluca** [comment]
> Yeah I'll need this too - fun fact, in order to be able to use fsverity the filesystem block size MUST be equal to the memory page size. So I'd even drop the 4k high limit, given it's not unlikely on arm64 to use larger pages.

I think I need more than 4k as an option? 16k pages on an arm64 is not uncommon

**Comment by DaanDeMeyer** [comment]
> I think I need more than 4k as an option? 16k pages on an arm64 is not uncommon

This is about filesystem block size, not sector size. We should probably add a section option to set the filesystem block size.

**Comment by bluca** [comment]
> > I think I need more than 4k as an option? 16k pages on an arm64 is not uncommon
> 
> This is about filesystem block size, not disk sector size. We should probably add a second option to set the filesystem block size.

Ah dang, I always get confused between the two

**Pull request merged**